File: User-Service/src/services/user.service.js
const { UserRepository } = require('../repositories');
const { comparePassword, validatePasswordStrength, hashPassword } = require('../utils/password.utils');
const { generateAccessToken, generateRefreshToken, verifyRefreshToken } = require('../utils/jwt.utils');
const BadRequest = require('../errors/badrequest.error');
const Unauthorized = require('../errors/unauthorized.error');
const NotFound = require('../errors/notfound.error');
const logger = require('../config/logger.config');

class UserService {
    constructor() {
        this.userRepository = new UserRepository();
    }

    /**
     * Register a new user
     * @param {Object} userData - User registration data
     * @returns {Promise<Object>} - User object without password
     */
    async registerUser(userData) {
        try {
            // Validate password strength
            const passwordValidation = validatePasswordStrength(userData.password);
            if (!passwordValidation.isValid) {
                throw new BadRequest('password', passwordValidation.errors.join(', '));
            }

            // Check if email already exists
            const existingUserByEmail = await this.userRepository.getUserByEmail(userData.email);
            if (existingUserByEmail) {
                throw new BadRequest('email', 'Email already registered');
            }

            // Check if username already exists
            const existingUserByUsername = await this.userRepository.getUserByUsername(userData.username);
            if (existingUserByUsername) {
                throw new BadRequest('username', 'Username already taken');
            }

            // Create user
            const user = await this.userRepository.createUser(userData);

            logger.info(`New user registered: ${user.email}`);

            // Return user without password
            return user.toJSON();
        } catch (error) {
            logger.error('UserService.registerUser:', error);
            throw error;
        }
    }

    /**
     * Login user
     * @param {string} email - User email
     * @param {string} password - User password
     * @returns {Promise<Object>} - Access token, refresh token, and user data
     */
    async loginUser(email, password) {
        try {
            // Find user by email
            const user = await this.userRepository.getUserByEmail(email);
            if (!user) {
                throw new Unauthorized('Invalid email or password');
            }

            // Check if user is active
            if (!user.isActive) {
                throw new Unauthorized('Account is deactivated');
            }

            // Verify password
            const isPasswordValid = await comparePassword(password, user.password);
            if (!isPasswordValid) {
                throw new Unauthorized('Invalid email or password');
            }

            // Generate tokens
            const payload = {
                id: user._id,
                email: user.email,
                username: user.username,
                role: user.role
            };

            const accessToken = generateAccessToken(payload);
            const refreshToken = generateRefreshToken(payload);

            logger.info(`User logged in: ${user.email}`);

            return {
                accessToken,
                refreshToken,
                user: user.toJSON()
            };
        } catch (error) {
            logger.error('UserService.loginUser:', error);
            throw error;
        }
    }

    /**
     * Refresh access token
     * @param {string} refreshToken - Refresh token
     * @returns {Promise<Object>} - New access token
     */
    async refreshAccessToken(refreshToken) {
        try {
            // Verify refresh token
            const decoded = verifyRefreshToken(refreshToken);

            // Get user from database
            const user = await this.userRepository.getUserById(decoded.id);

            if (!user.isActive) {
                throw new Unauthorized('Account is deactivated');
            }

            // Generate new access token
            const payload = {
                id: user._id,
                email: user.email,
                username: user.username,
                role: user.role
            };

            const newAccessToken = generateAccessToken(payload);

            return {
                accessToken: newAccessToken
            };
        } catch (error) {
            logger.error('UserService.refreshAccessToken:', error);
            throw error;
        }
    }

    /**
     * Get user profile
     * @param {string} userId - User ID
     * @returns {Promise<Object>} - User profile
     */
    async getUserProfile(userId) {
        try {
            const user = await this.userRepository.getUserById(userId);
            return user.toJSON();
        } catch (error) {
            logger.error('UserService.getUserProfile:', error);
            throw error;
        }
    }

    /**
     * Update user profile
     * @param {string} userId - User ID
     * @param {Object} updateData - Data to update (firstName, lastName)
     * @returns {Promise<Object>} - Updated user profile
     */
    async updateUserProfile(userId, updateData) {
        try {
            // Only allow updating specific fields
            const allowedUpdates = ['firstName', 'lastName'];
            const updates = {};

            for (const key of allowedUpdates) {
                if (updateData[key] !== undefined) {
                    updates[key] = updateData[key];
                }
            }

            if (Object.keys(updates).length === 0) {
                throw new BadRequest('updateData', 'No valid fields to update');
            }

            const updatedUser = await this.userRepository.updateUser(userId, updates);

            logger.info(`User profile updated: ${updatedUser.email}`);

            return updatedUser.toJSON();
        } catch (error) {
            logger.error('UserService.updateUserProfile:', error);
            throw error;
        }
    }

    /**
     * Change user password
     * @param {string} userId - User ID
     * @param {string} oldPassword - Current password
     * @param {string} newPassword - New password
     * @returns {Promise<Object>} - Success message
     */
    async changePassword(userId, oldPassword, newPassword) {
        try {
            // Get user
            const user = await this.userRepository.getUserById(userId);

            // Verify old password
            const isPasswordValid = await comparePassword(oldPassword, user.password);
            if (!isPasswordValid) {
                throw new Unauthorized('Current password is incorrect');
            }

            // Validate new password strength
            const passwordValidation = validatePasswordStrength(newPassword);
            if (!passwordValidation.isValid) {
                throw new BadRequest('newPassword', passwordValidation.errors.join(', '));
            }

            // Hash new password
            const hashedPassword = await hashPassword(newPassword);

            // Update password
            user.password = hashedPassword;
            await user.save();

            logger.info(`Password changed for user: ${user.email}`);

            return {
                message: 'Password changed successfully'
            };
        } catch (error) {
            logger.error('UserService.changePassword:', error);
            throw error;
        }
    }
}

module.exports = UserService;

File: User-Service/src/services/user_service.py
"""
User Service
Business logic for user management
"""
from typing import Optional, List
from src.repositories.user_repository import user_repository
from src.models.user_model import UserCreate, UserUpdate, UserResponse, UserRole
from src.utils.exceptions import NotFoundException, ConflictException
import logging

logger = logging.getLogger(__name__)


class UserService:
    """Service for user business logic"""
    
    def __init__(self):
        """Initialize the service with repository"""
        self.repository = user_repository
    
    async def get_user_by_id(self, user_id: str) -> UserResponse:
        """
        Get user by ID
        
        Args:
            user_id: User ID
            
        Returns:
            UserResponse model
            
        Raises:
            NotFoundException: If user not found
        """
        user = await self.repository.get_user_by_id(user_id)
        
        if not user:
            raise NotFoundException(f"User with ID {user_id} not found")
        
        return UserResponse(
            id=user["_id"],
            email=user["email"],
            username=user["username"],
            full_name=user.get("full_name"),
            role=user["role"],
            is_active=user["is_active"],
            is_verified=user["is_verified"],
            created_at=user["created_at"],
            last_login=user.get("last_login")
        )
    
    async def update_user_profile(self, user_id: str, update_data: UserUpdate) -> UserResponse:
        """
        Update user profile
        
        Args:
            user_id: User ID
            update_data: Data to update
            
        Returns:
            Updated UserResponse model
            
        Raises:
            NotFoundException: If user not found
            ConflictException: If email/username already exists
        """
        # Check if user exists
        user = await self.repository.get_user_by_id(user_id)
        if not user:
            raise NotFoundException(f"User with ID {user_id} not found")
        
        # Prepare update data
        update_dict = update_data.dict(exclude_unset=True)
        
        # Check for email conflict
        if "email" in update_dict:
            existing_user = await self.repository.get_user_by_email(update_dict["email"])
            if existing_user and existing_user["_id"] != user_id:
                raise ConflictException("Email already in use")
        
        # Check for username conflict
        if "username" in update_dict:
            existing_user = await self.repository.get_user_by_username(update_dict["username"])
            if existing_user and existing_user["_id"] != user_id:
                raise ConflictException("Username already in use")
        
        # Update user
        updated_user = await self.repository.update_user(user_id, update_dict)
        
        return UserResponse(
            id=updated_user["_id"],
            email=updated_user["email"],
            username=updated_user["username"],
            full_name=updated_user.get("full_name"),
            role=updated_user["role"],
            is_active=updated_user["is_active"],
            is_verified=updated_user["is_verified"],
            created_at=updated_user["created_at"],
            last_login=updated_user.get("last_login")
        )
    
    async def get_all_users(self, skip: int = 0, limit: int = 10) -> List[UserResponse]:
        """
        Get all users with pagination
        
        Args:
            skip: Number of records to skip
            limit: Maximum number of records to return
            
        Returns:
            List of UserResponse models
        """
        users = await self.repository.get_all_users(skip, limit)
        
        return [
            UserResponse(
                id=user["_id"],
                email=user["email"],
                username=user["username"],
                full_name=user.get("full_name"),
                role=user["role"],
                is_active=user["is_active"],
                is_verified=user["is_verified"],
                created_at=user["created_at"],
                last_login=user.get("last_login")
            )
            for user in users
        ]
    
    async def deactivate_user(self, user_id: str) -> bool:
        """
        Deactivate a user account
        
        Args:
            user_id: User ID
            
        Returns:
            True if successful
            
        Raises:
            NotFoundException: If user not found
        """
        user = await self.repository.get_user_by_id(user_id)
        if not user:
            raise NotFoundException(f"User with ID {user_id} not found")
        
        result = await self.repository.delete_user(user_id)
        
        if result:
            logger.info(f"User {user_id} deactivated")
            return True
        
        return False


# Global service instance
user_service = UserService()
